Develop

RescueRangerA opened the pull request.

a-mihalych committed:
feat: добавить удаление фильмов и пользователей (#5)

feat: добавить удаление фильмов и пользователей

Asror0314 committed:
Add director (#8)

Add director

RescueRangerA committed:
feat: add reviews (#6)

feat: add reviews

a-mihalych committed:
feat: добавить получение общих фильмов (#7)

feat: добавить получение общих фильмов

aoengovatov committed:
Add search (#9)

Добавление поиска Film по названию фильма и режиссеру

a-mihalych committed:
feat: добавить получение лучших фильмов по жанру и году (#10)

Co-authored-by: RescueRangerA <hammer_head@mail.ru>

aoengovatov committed:
Add feed (#12)

Add feed

RescueRangerA committed:
feat: add recommendations (#11)

RescueRangerA committed:
Develop pre fixes (#13)

* fix: duplicate key violation

* refactor: split the "film" request

* refactor: redundant variables

* fix: more tweaks

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/director/DirectorDbStorage.java:
```diff
@@ -0,0 +1,104 @@
+
+    private final JdbcTemplate jdbcTemplate;
+
+    public DirectorDbStorage(JdbcTemplate jdbcTemplate) {
```
Не хватает консистентности подхода - где-то есть @Autowired на конструкторе. где-то его нет. Оба варианта в целом верны, как и @RequiredArgsConstructor, только нужно придерживаться одного из них

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/director/DirectorDbStorage.java:
```diff
@@ -0,0 +1,104 @@
+        } catch (EmptyResultDataAccessException e) {
+        }
+
+        return director != null ? Optional.of(director) : Optional.empty();
```
(опционально) может просто Optional.ofNullable() ?

alexache commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmReviewController.java:
```diff
@@ -0,0 +1,73 @@
+    public FilmReview createFilmReview(@Valid @RequestBody FilmReview filmReview) {
+        filmReviewService.create(filmReview);
+
+        return getFilmReview(filmReview.getReviewId());
```
Выглядит как-то неоптимально, почему сразу не вернуть объект из .create()?

alexache commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmReviewController.java:
```diff
@@ -0,0 +1,73 @@
+    public FilmReview updateFilmReview(@Valid @RequestBody FilmReview filmReview) {
+        filmReviewService.update(filmReview);
+
+        return getFilmReview(filmReview.getReviewId());
```
И это тоже выглядит неоптимально

alexache commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmReviewService.java:
```diff
@@ -0,0 +1,138 @@
+        Optional<Film> film = filmStorage.findFilmById(filmReview.getFilmId());
+        Optional<User> user = userStorage.findById(filmReview.getUserId());
+
+        if (film.isEmpty()) {
```
(опционально) Можно бы делать .orElseThrow() вместо условий на Optional

alexache commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmReviewService.java:
```diff
@@ -0,0 +1,138 @@
+        }
+
+        filmReview = filmReviewStorage.saveFilmReview(filmReview);
+        filmReview = getById(filmReview.getReviewId());
```
Для чего эта операция?

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -217,4 +216,61 @@ public List<User> findFriendsInCommonOf2Users(User userEntityA, User userEntityB
+        Assert.notNull(rs.getString("operation"), "Operation must not be null.");
+        Enum<EventType> eventType = null;
+        switch (rs.getString("event_type")){
+            case "LIKE":
```
Может будет лучше воспользоваться чем-то вроде EventType.valueOf()

alexache commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -125,4 +159,45 @@ public FilmMpaRating findRatingById(Long aLong) {
+        List<Film> films = new ArrayList<>();
+        if(!query.isBlank() && !by.isBlank()){
+            String [] param = by.split(",");
+            if(param.length == 2){
```
Не помешает отформатировать с нужными пробелами и отступами

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -241,4 +237,267 @@ public List<Film> findTopNMostLikedFilms(Integer limit) {
+        Assert.notNull(limit, "Limit must not be null.");
+
+        if (limit < 1) {
+            return List.of();
```
Collections.emptyList() лучше подходит для таких случаев

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -241,4 +237,267 @@ public List<Film> findTopNMostLikedFilms(Integer limit) {
+        }
+
+        if ( bestUserId == null ) {
+            return List.of();
```
Collections.emptyList() лучше подходит для таких случаев

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -241,4 +237,267 @@ public List<Film> findTopNMostLikedFilms(Integer limit) {
+    }
+
+    protected List<Film> completeExternalEntitiesForFilms(List<Film> films) {
+        Map<Long, LinkedHashSet<Genre>> genresMap = buildGenresMap(films);
```
Лучше больше интерфейсов - Map<Long, Set<Genre>> genres
и не рекомендуется использовать тип в имени переменной (genresMap, genresSet и так далее)

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -241,4 +237,267 @@ public List<Film> findTopNMostLikedFilms(Integer limit) {
+        return films;
+    }
+
+    protected Map<Long, LinkedHashSet<Genre>> buildGenresMap(List<Film> films) {
```
Может просто private?

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -241,4 +237,267 @@ public List<Film> findTopNMostLikedFilms(Integer limit) {
+        );
+    }
+
+    protected Map<Long, LinkedHashSet<Director>> buildDirectorsMap(List<Film> films) {
```
Эти 2 метода немного пугают, но без них было бы хуже, и очень хорошо, что без запросов к базе в цикле

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/filmdirector/FilmDirectorDbStorage.java:
```diff
@@ -0,0 +1,48 @@
+    public Film saveDirectorsOfTheFilm(Film film) {
+        final String sqlQuery = "INSERT INTO film_director(film_id, director_id) VALUES (?,?)";
+
+        if(film.getDirectors().size() == 0) return film;
```
Не хватает пробелов для форматирования, да и не очень понятно, зачем этому методу возвращать объект фильма

RescueRangerA committed:
Develop tweaks (#15)

* refactor: formatting and logic

https://github.com/RescueRangerA/java-filmorate/pull/14#discussion_r1057358046

* refactor: change access modifier

https://github.com/RescueRangerA/java-filmorate/pull/14#discussion_r1057357728

* refactor: use Collections.emptyList() instead of List.of()

https://github.com/RescueRangerA/java-filmorate/pull/14#discussion_r1057357289
https://github.com/RescueRangerA/java-filmorate/pull/14#discussion_r1057357212

* refactor: formatting

https://github.com/RescueRangerA/java-filmorate/pull/14#discussion_r1057356618

* refactor: use valueOf() for enumerations

https://github.com/RescueRangerA/java-filmorate/pull/14#discussion_r1057355636

* refactor: use orElseThrow() instead of conditions

https://github.com/RescueRangerA/java-filmorate/pull/14#discussion_r1057353910

* refactor: remove redundant calls in controllers

https://github.com/RescueRangerA/java-filmorate/pull/14#discussion_r1057353244
https://github.com/RescueRangerA/java-filmorate/pull/14#discussion_r1057353281

* refactor: use Optional.ofNullable()

https://github.com/RescueRangerA/java-filmorate/pull/14#discussion_r1057352567

* refactor: missing autowired annotations

https://github.com/RescueRangerA/java-filmorate/pull/14#discussion_r1057352315

* refactor: use interface as type

https://github.com/RescueRangerA/java-filmorate/pull/14#discussion_r1057357626

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -379,23 +381,23 @@ public List<Film> searchByFilmAndDirector(String query) {
-            if (genresMap.containsKey(film.getId())) {
-                film.setGenres(genresMap.get(film.getId()));
+            if (genres.containsKey(film.getId())) {
+                film.setGenres((LinkedHashSet<Genre>) genres.get(film.getId()));
```
Как мы и обсудили, в идеале избавиться от таких кастов и наличия linkedHashSet из модели фильма
